engine/src/main/java/org/hibernate/search/indexes/impl/PeriodicRegreshingReaderProvider.java

```diff
@@ -0,0 +1,301 @@
+		this.scheduledExecutorService = Executors.newScheduledThreadPool( "Periodic IndexReader refreshing task for index " + indexName );
+		this.delay = ConfigurationParseHelper.getIntValue(
+				props,
+				Environment.ASYNC_READER_REFRESH_PERIOD_MS,
+				DEFAULT_ACCEPTABLE_MAX_AGE_MS );
+		this.scheduledExecutorService.scheduleWithFixedDelay(
+				new IndexRefreshTask(),
+				delay,
+				delay,
+				TimeUnit.MILLISECONDS
+		);
+	}
+
+	@Override
+	public void stop() {
+		if ( scheduledExecutorService != null ) {
+			try {
+				scheduledExecutorService.shutdown();
+				scheduledExecutorService.awaitTermination( 1, TimeUnit.SECONDS );
+				scheduledExecutorService = null;
+			}
+			catch (InterruptedException e) {
+				log.timedOutWaitingShutdownOfReaderProvider( indexName );
+			}
+		}
+		for ( IndexReader reader : allReaders.keySet() ) {
+			ReaderUsagePair usage = allReaders.get( reader );
+			usage.close();
+		}
+
+		if ( allReaders.size() != 0 ) {
+			log.readersNotProperlyClosedInReaderProvider();
+		}
+	}
+
+
+	/**
+	 * Thread safe creation of <code>PerDirectoryLatestReader</code>.
+	 *
+	 * @param directory The Lucene directory for which to create the reader.
+	 * @return either the cached instance for the specified <code>Directory</code> or a newly created one.
+	 * @see <a href="http://opensource.atlassian.com/projects/hibernate/browse/HSEARCH-250">HSEARCH-250</a>
+	 */
+	private synchronized PerDirectoryLatestReader createReader(Directory directory) {
+		PerDirectoryLatestReader reader = currentReaders.get( directory );
+		if ( reader != null ) {
+			return reader;
+		}
+
+		try {
+			reader = new PerDirectoryLatestReader( directory );
+			currentReaders.put( directory, reader );
+			return reader;
+		}
+		catch (IOException e) {
+			throw new SearchException( "Unable to open Lucene IndexReader for IndexManager " + this.indexName, e );
+		}
+	}
+
+
+	/**
+	 * Container for the couple IndexReader,UsageCounter.
+	 */
+	protected final class ReaderUsagePair {
+
+		public final DirectoryReader reader;
+		/**
+		 * When reaching 0 (always test on change) the reader should be really
+		 * closed and then discarded.
+		 * Starts at 2 because:
+		 * first usage token is artificial: means "current" is not to be closed (+1)
+		 * additionally when creating it will be used (+1)
+		 */
+		protected final AtomicInteger usageCounter = new AtomicInteger( 2 );
+
+		ReaderUsagePair(DirectoryReader r) {
+			reader = r;
+		}
+
+		/**
+		 * Closes the <code>IndexReader</code> if no other resource is using it
+		 * in which case the reference to this container will also be removed.
+		 */
+		public void close() {
+			int refCount = usageCounter.decrementAndGet();
+			if ( refCount == 0 ) {
+				ReaderUsagePair removed = allReaders.remove( reader );//remove ourself
+				try {
+					reader.close();
+				}
+				catch (IOException e) {
+					log.unableToCloseLuceneIndexReader( e );
+				}
+				assert removed != null;
```

Comment:
Any "current" IndexReader has an additional token in the `usageCounter`, so if close() method is invoked on the current IR, it won't actually be closed.

See also the comment on `usageCounter`.

Also on the method `refreshIfNeeded()` you'll see that the no-longer-current IndexReader is closed - yet this simply means that its usage counter is decremented, so that if any query is still using it it won't actually be closed, but it will only be closed when the last query which is using it will be closed.